Add I2C driver launcher

julianoes opened the pull request:
This is just a first draft of the launcher part.

Addresses https://github.com/PX4/PX4-Autopilot/issues/20681.

julianoes commented:
@davids5 any idea how I can get the `system()` call to work?

```
/home/julianoes/src/upstream/PX4-Autopilot/src/systemcmds/i2c_launcher/i2c_launcher.cpp:145:4: error: 'system' was not declared in this scope
  145 |    system(buf);
      |    ^~~~~~
```

julianoes commented:
Thanks @davids5. When I rename the systemcmd it is no longer built. How do I still get it to build from the px4board config?

davids5 commented:
> Thanks @davids5. When I rename the systemcmd it is no longer built. How do I still get it to build from the px4board config?

rerun make `px4_fmu-v6x boardconfig` and select the new name USER_SYSTEMCMDS_I2C_LAUNCHER

julianoes commented:
@davids5 I tried that but it doesn't build it anymore.

This is the diff I have and the `#error` is never hit.

```
diff --git a/boards/px4/fmu-v6x/default.px4board b/boards/px4/fmu-v6x/default.px4board
index 0d685d75cf..7ffe5b7565 100644
--- a/boards/px4/fmu-v6x/default.px4board
+++ b/boards/px4/fmu-v6x/default.px4board
@@ -79,8 +79,8 @@ CONFIG_SYSTEMCMDS_BSONDUMP=y
 CONFIG_SYSTEMCMDS_DMESG=y
 CONFIG_SYSTEMCMDS_GPIO=y
 CONFIG_SYSTEMCMDS_HARDFAULT_LOG=y
+CONFIG_USER_SYSTEMCMDS_I2C_LAUNCHER=y
 CONFIG_SYSTEMCMDS_I2CDETECT=y
-CONFIG_SYSTEMCMDS_I2C_LAUNCHER=y
 CONFIG_SYSTEMCMDS_LED_CONTROL=y
 CONFIG_SYSTEMCMDS_MFT=y
 CONFIG_SYSTEMCMDS_MTD=y
diff --git a/boards/px4/fmu-v6x/nuttx-config/nsh/defconfig b/boards/px4/fmu-v6x/nuttx-config/nsh/defconfig
index 13191fc716..1abc8370d7 100644
--- a/boards/px4/fmu-v6x/nuttx-config/nsh/defconfig
+++ b/boards/px4/fmu-v6x/nuttx-config/nsh/defconfig
@@ -205,7 +205,6 @@ CONFIG_SCHED_INSTRUMENTATION_SWITCH=y
 CONFIG_SCHED_LPWORK=y
 CONFIG_SCHED_LPWORKPRIORITY=50
 CONFIG_SCHED_LPWORKSTACKSIZE=1632
-CONFIG_SCHED_WAITPID=y
 CONFIG_SDMMC2_SDIO_PULLUP=y
 CONFIG_SEM_PREALLOCHOLDERS=32
 CONFIG_SERIAL_IFLOWCONTROL_WATERMARKS=y
@@ -285,6 +284,7 @@ CONFIG_SYSTEM_CDCACM=y
 CONFIG_SYSTEM_DHCPC_RENEW=y
 CONFIG_SYSTEM_NSH=y
 CONFIG_SYSTEM_PING=y
+CONFIG_SYSTEM_SYSTEM=y
 CONFIG_TASK_NAME_SIZE=24
 CONFIG_UART4_BAUD=57600
 CONFIG_UART4_RXBUFSIZE=600
diff --git a/src/systemcmds/i2c_launcher/Kconfig b/src/systemcmds/i2c_launcher/Kconfig
index b0d7de94ae..320fd69de2 100644
--- a/src/systemcmds/i2c_launcher/Kconfig
+++ b/src/systemcmds/i2c_launcher/Kconfig
@@ -1,4 +1,4 @@
-menuconfig SYSTEMCMDS_I2C_LAUNCHER
+menuconfig USER_SYSTEMCMDS_I2C_LAUNCHER
        bool "i2c_launcher"
        default n
        ---help---
diff --git a/src/systemcmds/i2c_launcher/i2c_launcher.cpp b/src/systemcmds/i2c_launcher/i2c_launcher.cpp
index aa02884959..e5d0dd0183 100644
--- a/src/systemcmds/i2c_launcher/i2c_launcher.cpp
+++ b/src/systemcmds/i2c_launcher/i2c_launcher.cpp
@@ -30,6 +30,7 @@
  * POSSIBILITY OF SUCH DAMAGE.
  *
  ****************************************************************************/
+#error
 
 #include <stdlib.h>
 #include "i2c_launcher.hpp"

```

davids5 commented:
@julianoes - well that is embarrassing...  @jlaitine How do we ensure __KERNEL__ is not set from https://github.com/PX4/PX4-Autopilot/pull/19080/commits/879b7c909bb36d6a9ab279b7d1e4329fc1afe7fa ?

jlaitine commented:
Hi,

That is a good question! There are several angles to this:

1. To avoid __KERNEL__ being set, the correct version in Kconfig would be:

```
config SYSTEMCMDS_I2C_LAUNCHER

...

config USER_I2C_LAUNCHER
	bool
	default 
	depends on SYSTEMCMDS_I2C_LAUNCHER

```

2. Interesting part is, whether "system" actually works in nuttx flat build, I have no idea?
3. I believe explicitly adding the __KERNEL__ compilation flag (or even populating the config_kernel_modules list) in "basic" CONFIG_BUILD_FLAT type builds might be just wrong. Maybe I should try to patch this?

jlaitine commented:
Could you check if it behaves differently with this patch?

https://github.com/PX4/PX4-Autopilot/pull/22063

julianoes commented:
Thanks @jlaitine!

So, I used your PR, and then:

```
 menuconfig SYSTEMCMDS_I2C_LAUNCHER
        bool "i2c_launcher"
        default n
        ---help---
                Daemon that starts drivers based on found I2C devices.

menuconfig USER_I2C_LAUNCHER
       bool
       default n
       depends on SYSTEMCMDS_I2C_LAUNCHER
```

And default.px4board:
```
CONFIG_USER_I2C_LAUNCHER=y
CONFIG_SYSTEMCMDS_I2CDETECT=y
```

And it worked!

julianoes commented:
@dagar this seems to be working for me! It would be good to get your review before I continue. And you can say what else you want done before merging this for 6x only.

dagar commented:
This looks like a good start, but we need to be skipping addresses that already have something running. We maintain a linked list of I2CSPIInstances. 

https://github.com/PX4/PX4-Autopilot/blob/c45eb29d1ab0c12926b89c1fe40ead1ef4705376/platforms/common/include/px4_platform_common/i2c_spi_buses.h#L110-L138

dagar commented on src/systemcmds/i2c_launcher/i2c_launcher.cpp:
```diff
@@ -0,0 +1,221 @@
+
+I2CLauncher::I2CLauncher() :
+	ModuleParams(nullptr),
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
```
From a scheduling and locking perspective this would be a bit cleaner if running from the relevant I2Cx work queue.

julianoes replied:
Hm, but that would be 4 queues, right?

dagar replied:
For whichever bus is relevant. Then from a scheduling perspective it should do something quick and minimal (eg check one address this cycle), then get out of the way for any real drivers already running.

julianoes replied:
Note that it doesn't run when armed.

dagar replied:
We still need to be careful not to disrupt any other drivers running on that bus prearm.

julianoes replied:
But if this runs for all external buses, it would schedule itself on multiple different work queues? Can you change the work queues on to easily like that?

dagar replied:
I would just keep it simple with one instance running per external bus, and only optimize that later if we need to.

julianoes replied:
Done.

julianoes commented:
@dagar I tried to check using the `BusInstanceIterator` but it seems to lock itself out somehow. Is that the right way?

```
	for (unsigned i = 0; i < sizeof(_devices) / sizeof(_devices[0]); ++i) {

		BusCLIArguments bus_cli_arguments{true, false};
		BusInstanceIterator i2c_bus_instance_iterator {"i2c_launcher", bus_cli_arguments, _devices[i].devid_driver_index};

		bool running = false;
		while (i2c_bus_instance_iterator.next()) {
			if (i2c_bus_instance_iterator.runningInstancesCount() > 0) {
				printf("%d -> running\n", i);
				running = true;
				break;
			} else {
				printf("%d -> not running\n", i);
			}
		}

		if (running) {
			continue;
		}
```

julianoes commented:
@dagar I need your help for the `BusInstanceIterator`, see above, thanks.

julianoes commented:
Tested on the bench with INA226, INA228, and INA238 on Pixhawk 6X.

@dagar I've addressed your review concerns. I think this should be good enough for a first version to be merged.

julianoes commented:
I've fixed an issue where the buses were not independent as they ought to be. @dagar check if you agree with https://github.com/PX4/PX4-Autopilot/pull/22047/commits/c957069db37d048f826422d4905214d17d34b8e1.

Rebased and force pushed.

dagar commented on platforms/common/i2c_spi_buses.cpp:
```diff
@@ -384,7 +384,7 @@ int BusInstanceIterator::runningInstancesCount() const
 
 	for (const auto &modules : i2c_spi_module_instances) {
-		if (strcmp(modules->_module_name, _module_name) == 0) {
+		if (modules->_bus == bus() && strcmp(modules->_module_name, _module_name) == 0) {
```
I think we don't want to break this so that the instance numbering is across all instances on all buses.

How does it break i2c launcher?

julianoes replied:
It shows modules running on other buses that we're not interested in. And `runningInstancesCount()` isn't used anywhere if I checked that right.

dagar replied:
Here: https://github.com/PX4/PX4-Autopilot/blob/c0609f3df6a2add0c63c7abc8875d01c2b7e2a38/platforms/common/i2c_spi_buses.cpp#L628

julianoes commented:
@dagar any objections here? Can we get this in?

julianoes commented:
@dagar last call...

julianoes commented:
@dagar I guess we only want to use this if the params like SENS_EN_INA226, SENS_EN_INA228, SENS_EN_INA238 are all set to 0? And then we can have the default for 6x so that all 3 set to 0. That way as soon as someone explicitly configures it, the launcher is no longer used.

Alternatively, we remove the 3 params and use the launcher exclusively.

julianoes commented:
> @dagar I guess we only want to use this if the params like SENS_EN_INA226, SENS_EN_INA228, SENS_EN_INA238 are all set to 0? And then we can have the default for 6x so that all 3 set to 0. That way as soon as someone explicitly configures it, the launcher is no longer used.

Ok, this is what I will implement.

julianoes commented:
@dagar done, using a separate method.
